colossalai/inference/core/engine.py

```diff
@@ -200,8 +217,76 @@ def clear_spec_dec(self) -> None:
         self.use_spec_dec = False
 
+    def convert_to_glide_model(
+        self, model: nn.Module, state_dict: Union[Dict, OrderedDict], strict: bool = True
+    ) -> nn.Module:
+        """
+        Convert the given model to a glide model.
+
+        Args:
+            model (nn.Module): The drafter model to be converted to GLIDE model.
+            state_dict (Union[Dict, OrderedDict]): The state dict to be loaded to the converted model.
+            strict (bool, optional): Whether to strictly load the state dict. Defaults to True.
+
+        Returns:
+            nn.Module: The converted glide model.
+        """
+        # get the policy corresponding to the drafter model from policy map
+        model_type = model.config.model_type
+        glide_type = f"glide_{model_type}"
+        if glide_type not in model_policy_map:
+            raise ValueError(f"GLIDE type {glide_type} is not supported yet. Please check the model type {model_type}")
+        policy = model_policy_map[glide_type]
+
+        # shard the drafter model add corresponding GLIDE layer
+        glide_model = self._shardformer(model, policy())
+        glide_model.load_state_dict(state_dict, strict=strict)
+
+        return glide_model
+
+    def glide_drafter_from_pretrained(self, model_path: Union[str, os.PathLike], config: PretrainedConfig) -> nn.Module:
+        """
+        Load and prepare a pretrained glide model used as a drafter model, from the given path.
+
+        Usage:
+        ```python
+        glide_config = GlideLlamaConfig(
+            intermediate_size=8192,
+            large_hidden_size=4096,
+            large_num_attention_heads=32,
+            num_hidden_layers=1,
+        )
+        # create a GLIDE drafte model
+        drafter_model = engine.glide_drafter_from_pretrained(drafter_model_path, glide_config)
+        ```
+
+        Args:
+            model_path (Union[str, os.PathLike]): The path to the pretrained glide model.
+            config: Glide model config.
+
+        Returns:
+            nn.Module: The model ready to be used as a GLIDE drafter model.
+        """
+        drafter_model = AutoModelForCausalLM.from_pretrained(config)
+        # For now, we try to support the same set of base models for glide models (drafter)
+        # as those for main models (verifier)
+        model_name = drafter_model.__class__.__name__
+        if model_name not in _supported_models:
+            raise ValueError(f"Model {model_name} is not supported yet as a glide drafter.")
+
+        # load params from the model path
+        files = [f for f in os.listdir(model_path) if f.endswith(".pth") or f.endswith(".pt") or f.endswith(".bin")]
+        # assume only use a single checkpoint file for drafter model
+        file_path = os.path.join(model_path, files[-1])
+        state_dict = torch.load(file_path)
+
+        drafter_model = self.convert_to_glide_model(drafter_model, state_dict)
+
+        return drafter_model
```

Comment:
We don't need to init the glide model inside engine, can move this part to an individual function.

Reply:
It actually uses the shard config constructed from the inference config of the engine and makes it easier for user to load the drafter model (For now TP is not supported yet though). Removing the utility func will make the loading process as follows: 

```python
glide_config = GlideLlamaConfig(
    intermediate_size=8192,
    large_hidden_size=4096,
    large_num_attention_heads=32,
    num_hidden_layers=1
)
drafter_model = LlamaForCausalLM(glide_config)
shard_config = ShardConfig()
shardformer = ShardFormer(shard_config=shard_config)
drafter_model, _ = shardformer.optimize(drafter_model, GlideLlamaModelPolicy())
state_dict = torch.load(checkpoint_path)
drafter_model.load_state_dict(state_dict)
```

Is it okay to let the user to shard and load the drafter model?

Reply:
Modified to use `GlideLlamaForCausalLM` in inference modeling, instead of replacing layers by ShardFormer.